Planner: add continuous strategy (BC)

iseeberg79 opened the pull request:
Fix https://github.com/evcc-io/evcc/issues/24461, fix https://github.com/evcc-io/evcc/issues/24769, fix https://github.com/evcc-io/evcc/issues/24486

**SinglePlan / Continuous Window Mode (optimized)**

- Creates a single continuous charging window for the required duration.
- Chooses the window with the lowest cost.
- preconditioning can be added at the end.
- Ensures uninterrupted charging.

**Cheapest Window / Default Mode (old)**

- Uses the cheapest combination of available slots to cover the required duration.
- Slots do not need to be continuous.


**Video**

[plan.webm](https://github.com/user-attachments/assets/55cee53e-9f67-4088-b0de-39c8d1608298)

**Screenshots**
<img width="870" height="651" alt="image" src="https://github.com/user-attachments/assets/04f677a4-ef6f-47b3-8e9c-f4999d479ac0" />

<img width="869" height="915" alt="image" src="https://github.com/user-attachments/assets/4a726983-4093-4509-980e-641524a73032" />

<img width="869" height="752" alt="image" src="https://github.com/user-attachments/assets/c77b927c-b663-4ce6-871b-38a472343670" />

<img width="869" height="752" alt="image" src="https://github.com/user-attachments/assets/01cbcf50-1c68-46b7-861f-43fcb4659af2" />

<img width="869" height="752" alt="image" src="https://github.com/user-attachments/assets/8506c884-3fe1-476e-bb48-9435cd6c453d" />

<img width="869" height="915" alt="image" src="https://github.com/user-attachments/assets/e48c0e8b-4f89-4999-8b1c-1d174dba61cb" />


**Breaking Change**

The precondition / late charging feature has moved from plan level to vehicle level. This simplifies our API and UI but means that all plans of a vehicle (static&repeating) use the same precondition option now.

TODOs
- [x] validate performance
- [x] validate results
- [x] validate UI 
- [x] validate repeating plans UI (long text)
- [x] validate repeating plans function
- [x] validate integration (soc-based, energy based, plan execution)
- [x] fix tests: mock
- [x] check for persisted plan settings
- [x] fix tests: integration (combobox)
- [x] double-check playwright adaptions
- [x] decide smallSlotDuration/smallGapDuration (and verify)
- [x] document api changes in `openapi.yaml`
- [ ] double-check commit (e8fd4c3)
- [ ] impact on integrators because of planner API changes

iseeberg79 committed:
UI adaption

UI integration

iseeberg79 committed:
new algorithm for single window

iseeberg79 committed:
backend integration

iseeberg79 committed:
add QUARTER_HOUR

iseeberg79 committed:
linter

iseeberg79 committed:
add test

iseeberg79 committed:
fix bug

testing has nil

iseeberg79 committed:
fix semantic error (costs)

naltatis commented:
> I'd love the flexibility in the UI. What about reusing the late toggle for advanced options, defaulting to 0min preconditioning and the option to enable single slot planning?

Interesting idea. Let me think on this.

> A loadpoint setting would define a planning default which would be a great extension

I wouldn't want to go the route of having configurable defaults and per item overwrites. That's indeed most flexible but also a hard to understand and visualize mechanism.

iseeberg79 committed:
completed backend integration

iseeberg79 committed:
correct mocking errors

adapt MockAPI

iseeberg79 committed:
happy linter

iseeberg79 committed:
happy UI linter

iseeberg79 commented:
I read your comment once more, thought about and absolutely agree, top level configuration of planning strategy that applies to any repeating plan makes sense and simplifies.

iseeberg79 commented:
What I was thinking about: is it possible and useful to somehow integrate or visualize the loadpoint gap constraints in the UI (preview)? I'd love to try applying some kind of mask for the constraints to get visible (preview level at least). What do you think?

naltatis commented:
@iseeberg79 can you explain, what you mean with "visualize the loadpoint gap constraints" exactly?

iseeberg79 committed:
fix test

time.Hour is problematic on new tests, same behavor on old test cases

iseeberg79 committed:
fix sliding windows optimization

iseeberg79 committed:
mock assets

iseeberg79 committed:
fix UI test

naltatis committed:
move continuous/precondition from plan to vehicle/lp; introduce plan strategy

naltatis committed:
simplify: remove user-selectable preview for repeating plans

naltatis committed:
add ui

naltatis committed:
lint

iseeberg79 committed:
fix continuous plans selected in the past

added tests

iseeberg79 committed:
handling for loadpoint

!soc-based plans

iseeberg79 committed:
performance

iseeberg79 committed:
fix sync values with backend

iseeberg79 committed:
consistent renaming

using effectivePlanPrecondition and effectivePlanContinuous

iseeberg79 committed:
fix reentrant lock

iseeberg79 committed:
happy linter

iseeberg79 commented:
> UI is still work in progress. haven't looked at the failing test cases yet but the manual tests I did looked good enough to push :)

they do! have done some corrections - hope not to interfere, revert otherwise.
Integration needs a new mock case as the precondition is tested and got re-designed.

iseeberg79 committed:
simplify

iseeberg79 committed:
prefer latest cheapest window

iseeberg79 committed:
fix return value

instead of plan-id, return soc

iseeberg79 committed:
add test ids

iseeberg79 committed:
fix playwright

to be double-checked

iseeberg79 commented:
looks good to me now

Is the missing combobox of "repeating › preview static and repeating plan" intentional? At the moment the combobox is missing and the test got adapted. The preview plan isn't selectable now and shows the latest plan which isn't intended, or? I guess it's part of the "UI in progress"..

<img width="832" height="836" alt="image" src="https://github.com/user-attachments/assets/89f6b4ac-feda-4e50-8750-6c53d984cacf" />

iseeberg79 committed:
Revert renaming

revert fix sync values with backend
revert consistent renaming

iseeberg79 committed:
sync on start with backend, avoid race-condition on change

iseeberg79 committed:
happy linter

iseeberg79 commented:
OK. 

What about smallGap/smallSlot? 

Agreed on: revert to skip small single slot shorter than 10minutes

But change also to to:
Allow gaps of <15min (instead of 30/60 minutes) to be more agressive on not-charging during skipped slots by planner. 
As it's not intended to charge between planner windows if for example up-to three slots have a high cost?

This would break today's "protection function" and is probably too aggressive for some APIs already having trouble on skipping >60m slots in the past? On the other hand a change to 15 is straight forward on slotduration change from 60 to 15 minutes. It will still allow charge at least during one full planner skipped slot (15min duration)?

A change of those should be done in another commit as it's unrelated to changes here?

naltatis commented:
> The preview plan isn't selectable now and shows the next active plan which isn't intended, or?

Yes, I removed the manual preview selection intentionally. Users can still preview by enabling/disabling specific plans. There was a lot of code (we'd also have to adapt) surrounding this preview features. Also I did not want two action areas (underlined text + button) so close together, since the user could confuse them to be one control.

Yes, haven't adapted the test cases. I'll review your changes later. Thanks!

iseeberg79 commented:
Yes, making the UI planner results getting exactly used.

iseeberg79 commented:
If a higher tolerance than slotDuration would be required, it should be shown by the planner UI. That could by done by applying kind of "planner mask" - to stick to showing executable plans only. In a separate PR if really needed.

andig commented:
s/günstig/günstigst. Günstig ist es ja immer!

iseeberg79 commented:
Ich find "optimal" auch zutreffend, aber das ist relativ zur Perspektive 
"günstigst." liest sich schwerer als "cheapest"; vielleicht "minimal"?

iseeberg79 committed:
i18n adaptions

iseeberg79 commented:
> But make sure shorter, not shorter or as short as.

Might the control cycle be relevant? Something like following formula: 

remainingWindowDuration + (controlCyleInterval + someProcessingTime) >= slotDuration

It's the reason why 10min would be better than todays 15min or "shorter than slotDuration". Equivalent would be greater than a portion of slotDuration, but ensure to be not lower than "x" for some minimum on very short slotDuration?

Did not follow your progress in the referenced follow-up in detail, yet.
So perhaps this thoughts are already outdated and are meanwhile offtopic here.

andig committed:
Replace small slot/gap durations with 15min standard slot duration

iseeberg79 commented on core/loadpoint_plan.go:
```diff
@@ -165,7 +161,7 @@ func (lp *Loadpoint) plannerActive() (active bool) {
 
 	if active {
 		// ignore short plans if not already active
-		if slotRemaining := lp.clock.Until(activeSlot.End); !lp.planActive && slotRemaining < smallSlotDuration && !planner.SlotHasSuccessor(activeSlot, plan) {
```
lp.clock.until(SlotEnd) will (almost) ever be shorter than slotDuration on singleSlots; before it was comparing against 10 minutes, recently against 15 minutes - not a problem for multiple spanning slots as slotRemaining is a multiple of slotDuration (>1).

iseeberg79 commented on core/loadpoint_plan.go:
```diff
@@ -165,7 +161,7 @@ func (lp *Loadpoint) plannerActive() (active bool) {
 	if active {
 		// ignore short plans if not already active
-		if slotRemaining := lp.clock.Until(activeSlot.End); !lp.planActive && slotRemaining < smallSlotDuration && !planner.SlotHasSuccessor(activeSlot, plan) {
+		if slotRemaining := lp.clock.Until(activeSlot.End); !lp.planActive && slotRemaining < tariff.SlotDuration && !planner.SlotHasSuccessor(activeSlot, plan) {
```
lp.clock.until(SlotEnd) will (almost) ever be shorter than slotDuration on singleSlots; before it was comparing against 10 minutes, recently against 15 minutes - not a problem for multiple spanning slots as slotRemaining is a multiple of slotDuration (>1).

iseeberg79 commented on core/loadpoint_plan.go:
```diff
@@ -184,11 +180,11 @@ func (lp *Loadpoint) plannerActive() (active bool) {
 			return true
-		case lp.clock.Until(planStart) < smallGapDuration:
-			lp.log.DEBUG.Printf("plan: avoid re-start within %v, continuing for remaining %v", smallGapDuration, lp.clock.Until(planStart).Round(time.Second))
+		case lp.clock.Until(planStart) < tariff.SlotDuration:
```
potentially the same because of until here as well?

iseeberg79 reviewed (commented):
> > I expect a message like "slot duration too short" because it's testing 00:14:59 against 00:15:00
> 
> What/ where/ why? A 15min slot is not a 14:59 slot.

root-cause?

iseeberg79 commented:
I am not able to point it out. Double-checked; logic 100% compatible and effective! Was not able to reproduce any issues - short slots are executed and gap continued as expected by the user and shown in UI. Wrote tests to verify - all passed.

naltatis committed:
Merge branch 'master' into feature/planner-15m-singleplan

naltatis committed:
improve tests; a11y; i18n structure; small ui fixes

naltatis committed:
precondition deprecation

naltatis committed:
update openapi

naltatis committed:
readd missing key

naltatis reviewed (approved):
Looks good from my side. The change is quite significant and the scenarios are not trivial - despite our test suite. So we have to keep an eye on unexpected behavior after rolling this out.

@andig last check?

naltatis committed:
gen mcp

iseeberg79 committed:
adjust preconditioning and cost-calculation handling

fix #24637

removed merging continuous slots

iseeberg79 committed:
fix

andig commented:
Something's broken. When I add late charging 30min, the wrong number of slots are added:

<img width="881" height="415" alt="Screenshot 2025-10-23 at 11 26 11" src="https://github.com/user-attachments/assets/9ce8220c-30c2-4085-8382-7a695ab6b057" />

Reverting to "none" will still show late slots.

andig commented:
...and plan display doesn't match plan duration:
<img width="837" height="682" alt="Screenshot 2025-10-23 at 11 28 01" src="https://github.com/user-attachments/assets/3fb2d64e-36fd-48ba-8311-6b56ef0566db" />

iseeberg79 commented:
> Something's broken. When I add late charging 30min, the wrong number of slots are added:
> 
> 
> 
> <img width="881" height="415" alt="Screenshot 2025-10-23 at 11 26 11" src="https://github.com/user-attachments/assets/9ce8220c-30c2-4085-8382-7a695ab6b057" />
> 
> 
> 
> Reverting to "none" will still show late slots.

@naltatis timing problem, realized shortly as well. if repeated it works (with changed values). I think there is a problem backend/frontend timing. Tried a fix - but still is happening. Could you help? 

@andig what's the target time? Did not happen with old implementation. Will try to fix forward or revert.

andig committed:
Simplfy- remove all duplicated code

andig committed:
More deduplication

andig commented:
Does the UI even fetch the new plan after updating the strategy @naltatis ?

andig committed:
wip

andig commented on core/site_vehicles.go:
```diff
@@ -41,26 +44,29 @@ func (site *Site) publishVehicles() {
+			MaxCurrent:       ac.MaxCurrent,
+			Priority:         ac.Priority,
+			Features:         lo.Map(instance.Features(), func(f api.Feature, _ int) string { return f.String() }),
+			Plan:             plan,
```
All these should be moved in to a sub-sctruct

andig commented on core/vehicle/adapter.go:
```diff
@@ -144,3 +139,20 @@ func (v *adapter) GetRepeatingPlans() []api.RepeatingPlanStruct {
+func (v *adapter) SetPlanStrategy(planStrategy api.PlanStrategy) error {
+	v.log.DEBUG.Printf("update plan strategy for vehicle %s (precondition: %vs, continuous: %v)", v.name, planStrategy.Continuous, planStrategy.Precondition)
+
+	settings.SetJson(v.key()+keys.PlanStrategy, planStrategy)
```
handle error

andig commented:
I would expect finding a continuous plan to be as simple as:

```go
func (t *Planner) continuousPlanV2(rates api.Rates, effectiveDuration time.Duration, targetTime time.Time) (api.Rates, float64) {
	slots := int(math.Ceil(float64(effectiveDuration) / float64(tariff.SlotDuration)))

	bestSlot := -1
	bestCost := math.MaxFloat64

	for i := range rates {
		lastSlot := i + slots - 1
		if !rates[lastSlot].Start.Before(targetTime) {
			break
		}

		var sum float64
		for j := i; j <= lastSlot; j++ {
			sum += rates[j].Value
		}

		// prefer later start if equal cost
		if sum <= bestCost {
			bestSlot = i
			bestCost = sum
		}
	}

	if bestSlot < 0 {
		return nil, 0
	}

	return rates[bestSlot : bestSlot+slots], bestCost
}
```

Avoids any intermediate data generation or collection incl. expensive copying of plans.

iseeberg79 committed:
adjust tests

iseeberg79 commented:
double checked the tests, some adapted to reflect API changes

Also introduced the missing marked TODO to test plans where rates might start in the future.

@andig I cannot reproduce your results with soc or energy based charging. What here sometimes happens is that made selection aren't effective: UI shows change of parameter, but backend does not update the planner reflecting the changed selected parameters - I assume a timing problem as I wrote before.

As I cannot reproduce the issue it's difficult to write a test. I'd like to go for the simplification next.

andig commented:
I‘ll reproduce tomorrow and also look at the api requests to get a better idea. I don‘t believe in timing since everything is synchronous.

iseeberg79 committed:
adapted to 15m logic (short early/late precondition)

iseeberg79 committed:
intregrate simpler approach

iseeberg79 committed:
fix goal not reached on precondition

iseeberg79 committed:
trim & costs

iseeberg79 committed:
condense

iseeberg79 committed:
tests and fix

andig commented:
Still happens, only with "continuous" and not always. In continuous mode, changing the precon duration will not always show up (correctly) in the ui. The strategy update is correctly applied, but the plan preview seems to deliver the wrong plan.

iseeberg79 commented:
Strategy and Precondition are getting persisted, they need to get passed for the preview and on activation stored as well. Something must get handled wrong in this area. 

So it's perhaps 

- persisting the value only on activation of plans
- http handler for the preview
- how the plan gets the value passed on preview/during operation

relevant to check.

I'll investigate later but I am not very experienced on. @naltatis what do you think - right assumption? Could you help?

andig committed:
Debug: round-trip strategy

andig commented:
Gefühlt geht es immer dann schief, wenn man vor Änderung der precon Dauer auch die Strategie ändert. Ich finde allerdings kein Muster. Der preview request zeigt, dass die falsche Strategie zurück gegeben wird.

andig committed:
Error handling, deduplicate

andig committed:
Reuse

andig committed:
wip

andig commented:
@naltatis können wir beim `addRepeatingPlansHandler` auf das wrapper struct verzichten? Json kann ja auch direkt Arrays zurück geben? Minimal eleganter...

andig committed:
wip

andig committed:
wip

iseeberg79 committed:
publish effective values

iseeberg79 committed:
fix nil exception

iseeberg79 committed:
Revert "publish effective values"

This reverts commit 5f62872e36444e336ed1f4687000ce993a90fe60.

iseeberg79 committed:
non-blocking

browser is pending a long time without

iseeberg79 committed:
persist after apply

andig commented:
Hier müssten doch 2 Dinge klar sein: preview darf nichts speichern und Frontend darf die Previewdaten nicht aus dem Websocket holen sondern nur übers API.

iseeberg79 committed:
simplify and add test

continous slot boundary is actually not reaching target on target time within slots (ie. 03:10)

iseeberg79 commented:
@naltatis could you review the changes? 
preview is not persisting strategy on change, soc-based and energy-based look good

@andig In the meantime, I had a situation where browser requests were blocked while the strategy was being saved (short time, potential for deadlocks?) due to lp.publishEffectiveValues (context: vehicle). I tried to analyze this, but didn't fully understand what was going on there. Not saving the strategy adjustment in the UI immediately already helps here, but could you please take a look at it? I made it asynchronous for now (commit e8fd4c397). It also works now without it! During the implementation attempts, which I have already undone, I got stuck there because the use of lp.publishEffectiveValues led to a complete blockage of requests (race condition) in various situations when the strategy was to be saved. Currently, this is only used in one place in the existing code.
As aynchronous is probably sematically unwanted, we might revert that change (it's isolated either).

<img width="491" height="863" alt="image" src="https://github.com/user-attachments/assets/7c15a8a0-8a00-451e-b82e-d1a89c3c9152" />

andig commented on core/planner/planner.go:
```diff
@@ -195,64 +262,114 @@ func (t *Planner) Plan(requiredDuration, precondition time.Duration, targetTime
+	var plan api.Rates
+	if requiredDuration > 0 {
+		if continuous {
+			// check if available rates span is sufficient for sliding window
```
Wird das nicht eingangs shcon geprüft? Warum brauchts das nur für sliding window? auch cheapest braucht doch exakt genau so viele Slots?

iseeberg79 replied:
wird nicht vorher geprüft, wird hier gebraucht um nicht zu planen (requiredDuration == 0) und zum Schluß bei Bedarf nur die Precondition anzufügen.

andig commented on core/planner/planner.go:
```diff
@@ -195,64 +262,114 @@ func (t *Planner) Plan(requiredDuration, precondition time.Duration, targetTime
 
-func splitPreconditionSlots(rates api.Rates, precondition time.Duration, targetTime time.Time) (api.Rates, api.Rates) {
-	var res, adjusted api.Rates
+func splitAndAdjustPrecondition(rates api.Rates, preCondStart time.Time, precondition time.Duration, targetTime time.Time) (api.Rates, api.Rates) {
```
Es wäre hilfreich, hier master zu mergen damit man sehen kann, was davon wirklich diff ist.

iseeberg79 commented:
> Everything ist kaputt. 
auf dem Handy hab ich einen Schreck bekommen.. "alles kaputt" ;-)
schaue ich mir gerade an, liegt am neuen precondition-Handling, fix kommt

iseeberg79 committed:
fix everything

iseeberg79 committed:
fix edge case

iseeberg79 committed:
clean-up tests

iseeberg79 committed:
merge master into branch

* chore: remove invalid false require definitions (#24625)

* Hardy-Barth Salia: enable new api at 2.3.64 (#24636)

* Fronius GEN24: document ECA (#24653)

* Calc plugin: add min/max (#24652)

* Config UI: fix grid title (#24661)

* chore: escape doc templates (#24660)

* Optimizer: handle smart cost limit (#24655)

* chore: reduce database writes (#24672)

* chore: minor

* Planner: fix missing costs for precondition slots (#24645)

* chore: reduce amount of optimizer requests in case of error

* chore: lock optimizer against piling goroutines

* chore: add pprof

* Tibber: fix formula not used

* docs: remove telemetry get endpoint (#24691)

* InfluxDB: add meter titles (#24610)

* FoxEss H3 Smart: fix power (#24626)

* Messaging: resolve pointer values

* OpenWB Pro: add wakeup (cp interrupt) (#24706)

* cli/token: fix database vehicles ignored (#24716)

* chore: replace with AGENTS.md

* Wattsonic Gen3: add min/max soc (#24720)

* chore: add missing api endpoints (#24721)

* chore: simplify repeating plans (#24719)

* resolve merge conflicts

iseeberg79 committed:
Merge branch 'master' into feature/planner-15m-singleplan

iseeberg79 committed:
resolve

iseeberg79 committed:
fix repeating plans test

iseeberg79 committed:
de-duplicate

errors from merge

andig committed:
Reduce diff

andig committed:
maintain order

andig committed:
wip

andig committed:
wip

andig commented:
@iseeberg79 hast Du ein klares Bild, an welcher Stelle wir im Planner die verfügbaren Raten prüfen (also gibt es genug am Ende und ab sofort)? Interessiert uns nur der erste Fall (wär für mich ok)? Lässt sich das vielleicht einmalig machen bevor es in billing/kontinuierlich rein geht? Haben wir einen Test dafür?

iseeberg79 commented:
Ja, da hab ich geprüft und es gibt mehrere Tests in die Richtung.

andig committed:
simplify

andig committed:
Add TODOs

iseeberg79 committed:
remove redundancy

iseeberg79 committed:
Merge branch 'master' into feature/planner-15m-singleplan

iseeberg79 committed:
resolve merge conflict, re-generate assets

repeating plan preview refactored

naltatis commented:
> Strategie mit Preview mit übergeben, ebenso wie bei setzen des Planes? Nicht separat! Das macht nur zusammen Sinn…

Strategy changes are immediately saved to the loadpoint/vehicle.
Our preview endpoints are loadpoint-specific, so there is no need to pass strategy settings. Preview should use current loadpoint(/vehicle) strategy settings when producing the preview results.

I'll double check the implementation.

iseeberg79 commented:
Actually preview allows not storing changed strategy, should be fine? Was not before, UI is consistent as preview does not persist values until applied. But yes, asked for double check exactly that changes

naltatis committed:
Revert "persist after apply"

This reverts commit 557c490d45b2222fee7328ffc9ffff4dccbbc962.

naltatis committed:
revert

naltatis committed:
immediate update; simplify

naltatis commented:
@iseeberg79 I reverted your UI changes and made a few adjustments. For strategy it's imho ok to save immediately, even in a preview scenario. I agree that this might not be 100% clean from a content hierarchy perspective (1. preview headline, 2. auto-save strategy form), but this is a far simpler approach: code- and UI-complexity-wise. The user gets instant feedback for his actions in the chart below. So the auto-save should be obvious. Let's run with this and watch for feedback.

In my test cases changing strategy now immediately saves again and refreshes the plan/preview as expected.

⚠️ During my tests I encountered a situation where the number of shown slots returned by the API kept growing. Just by changing strategy and refreshing. They also did not match the number of expected slots (far to many). But I'm now unable to reproduce this situation. Maybe it was an in-between situation. 🤷

naltatis commented:
@iseeberg79 @andig can one of you resolve the conflict?

iseeberg79 committed:
Merge branch 'master' into feature/planner-15m-singleplan

iseeberg79 committed:
fix merge conflicts, simplify

iseeberg79 commented:
@naltatis I am able to reproduce - at least sometimes, need to double check!

iseeberg79 committed:
correct and simplify

iseeberg79 committed:
consistent timezone handling

iseeberg79 committed:
Merge branch 'master' into feature/planner-15m-singleplan

iseeberg79 committed:
fix merge conflict

naltatis committed:
Merge branch 'master' into feature/planner-15m-singleplan

andig committed:
wip

andig committed:
wip

andig commented on core/planner/planner.go:
```diff
@@ -31,44 +32,54 @@ func New(log *util.Logger, tariff api.Tariff, opt ...func(t *Planner)) *Planner
-	var plan api.Rates
+// filterRates filters rates to the given time window and optionally adjusts boundary slots
+func filterRates(rates api.Rates, start, end time.Time) api.Rates {
+	if len(rates) == 0 {
```
Not needed, remove

andig commented on core/planner/planner_test.go:
```diff
@@ -57,7 +48,9 @@ func TestPlan(t *testing.T) {
-		plan := p.plan(rates, time.Hour, clock.Now())
+		// filter rates to [now, now] window - should return empty
+		filteredRates := filterRates(rates, clock.Now(), clock.Now())
+		plan := p.plan(filteredRates, time.Hour, clock.Now())
```
```suggestion
		rr := filterRates(rates, clock.Now(), clock.Now())
		plan := p.plan(rr, time.Hour, clock.Now())
```

iseeberg79 committed:
wip

iseeberg79 committed:
revert timezone handling

iseeberg79 committed:
simplify now handling

iseeberg79 committed:
wip

iseeberg79 committed:
simplify validation

iseeberg79 committed:
inline

iseeberg79 committed:
Merge branch 'master' into feature/planner-15m-singleplan

iseeberg79 committed:
Merge branch 'feature/planner-15m-singleplan' of https://github.com/iseeberg79/evcc into feature/planner-15m-singleplan

iseeberg79 committed:
reduce comment

iseeberg79 committed:
remove redundant tests

iseeberg79 committed:
performance

iseeberg79 committed:
Merge branch 'evcc-io:master' into feature/planner-15m-singleplan

iseeberg79 committed:
remove redundant panic

RTTTC commented:
Also - is there a way to visualize the ignored slots and gaps in UI Plan ?
If it is shown as planned, but will be ignored anyway - that only makes it confusing and false.

iseeberg79 committed:
reduce noise

andig commented on core/planner/helper.go:
```diff
@@ -32,7 +32,10 @@ func Duration(plan api.Rates) time.Duration {
 	var duration time.Duration
 	for _, slot := range plan {
 		slotDuration := slot.End.Sub(slot.Start)
-		duration += slotDuration
```
Why if this can't happen? If it can happen it needs a test.

andig committed:
Reduce diff

andig commented on core/planner/planner_test.go:
```diff
@@ -141,7 +134,7 @@ func TestNilTariff(t *testing.T) {
 		clock: clock,
 	}
 
-	plan := p.Plan(time.Hour, 0, clock.Now().Add(30*time.Minute))
```
Macht's nicht einfacher zu lesen wenn sich bei jedem Aufruf die Reihenfolge der Parameter ändern ;)

iseeberg79 replied:
ne, stimmt ...

andig committed:
reorder

andig committed:
wip

andig committed:
wip

andig committed:
wip

iseeberg79 commented on core/planner/planner.go:
```diff
@@ -31,86 +32,146 @@ func New(log *util.Logger, tariff api.Tariff, opt ...func(t *Planner)) *Planner
 }
 
+// filterRates filters rates to the given time window and adjusts boundary slots
+func filterRates(rates api.Rates, start, end time.Time) api.Rates {
```
yes, thought about and agree

iseeberg79 committed:
renaming

filterRates renamed to clampRates

iseeberg79 committed:
remove defensive check

iseeberg79 committed:
remove defensive check

iseeberg79 committed:
prepare merge

iseeberg79 committed:
Merge branch 'master' into feature/planner-15m-singleplan

iseeberg79 committed:
fix

iseeberg79 committed:
Merge branch 'master' into feature/planner-15m-singleplan

iseeberg79 committed:
linter

andig commented on core/keys/loadpoint.go:
```diff
@@ -76,11 +79,13 @@ const (
 	PlanProjectedStart = "planProjectedStart" // charge plan start time (earliest slot)
 	PlanProjectedEnd   = "planProjectedEnd"   // charge plan ends (end of last slot)
 	PlanOverrun        = "planOverrun"        // charge plan goal not reachable in time
+	PlanStrategy       = "planStrategy"       // charge plan strategy (precondition, continuous)
```
ping @iseeberg79

andig replied:
Nvm. Allerdings ist die Inkonsistenz intern/extern irritierend ;)

andig committed:
Simplify

andig commented on core/planner/planner.go:
```diff
@@ -32,61 +32,41 @@ func New(log *util.Logger, tariff api.Tariff, opt ...func(t *Planner)) *Planner
 		// slot covers more than we need, so shorten it
-		if requiredDuration < 0 {
-			// the first (if not single) slot should start as late as possible
+		if slotDuration > requiredDuration {
```
I'm trying to revert this section to the original form:

```go
	for _, slot := range rates {
		slotDuration := slot.End.Sub(slot.Start)
		requiredDuration -= slotDuration

		// slot covers more than we need, so shorten it
		if requiredDuration < 0 {
			if IsFirst(slot, plan) && len(plan) > 0 {
				slot.Start = slot.Start.Add(slotDuration - requiredDuration)
			} else {
				slot.End = slot.End.Add(requiredDuration - slotDuration)
			}
			requiredDuration = 0
		}

		plan = append(plan, slot)

		// we found all necessary slots
		if requiredDuration == 0 {
			break
		}
	}
```

But tests keep failing. I don't get why- these two forms should be equivalant.

andig committed:
wip

andig committed:
Move clampRates

andig committed:
Merge branch 'master' into feature/planner-15m-singleplan

andig commented on core/planner/planner.go:
```diff
@@ -192,72 +143,96 @@ func (t *Planner) Plan(requiredDuration, precondition time.Duration, targetTime
+		t.log.DEBUG.Printf("planner: target time beyond available slots- reducing plan horizon from %v to %v",
 			requiredDuration.Round(time.Second), durationAfterRates.Round(time.Second))
 
+		// Time after rates will be charged anyway, covers the precondition requirement
```
Kommentar unklar. What covers what?

iseeberg79 replied:
hier geht's um nötige Ladezeit, die über die Tarifdaten hinaus geht - Test "TargetAfterKnownPrices"; wird geändert zu 
// TargetAfterKnownPrices: reduce requirements as charging after rates is assumed

andig committed:
wip

andig commented on core/planner/planner.go:
```diff
@@ -203,61 +155,84 @@ func (t *Planner) Plan(requiredDuration, precondition time.Duration, targetTime
+		total += p.End.Sub(p.Start)
+	}
+
+	if deficit := precondition - total; deficit > 0 {
```
Was bedeutet deficit?

andig replied:
Code muss noch entfernt werden oder braucht einen Test

iseeberg79 replied:
unter der Annahme vollständiger bzw. lückenloser Tarifdaten ist das ein edge-case: remove

andig committed:
wip

andig commented on core/planner/planner.go:
```diff
@@ -203,29 +155,55 @@ func (t *Planner) Plan(requiredDuration, precondition time.Duration, targetTime
 
+			// available window too small for sliding window - charge continuously from now to target
+			if end.Sub(start) < requiredDuration {
+				return continuousPlan(append(rates, precond...), now, targetTime.Add(precondition))
```
Isn't that just

    return continuousPlan(rates, latestStart, targetTime)

? Wondering if that should have been covered above. How do we end up here?

andig committed:
wip

andig committed:
wip

andig committed:
Simplify continuous

andig commented on core/planner/planner.go:
```diff
@@ -203,22 +155,51 @@ func (t *Planner) Plan(requiredDuration, precondition time.Duration, targetTime
+		}
+
+		// find cheapest continuous window
+		plan = findContinuousWindow(rates, requiredDuration, targetTime)
```
what happens if nil gets returned here? can this happen?

iseeberg79 replied:
If rates exist but none are usable, should we fall back to simplePlan (charge without optimization) instead of nil (don't charge)?

Current behavior is to potentionally missing the target. We could fall back to simplePlan for this edge case? But it should normally only happen if there is a problem with the tariff data

andig commented on core/planner/planner_single_test.go:
```diff
@@ -0,0 +1,260 @@
+	"go.uber.org/mock/gomock"
+)
+
+func TestContinuous_SinglePlanWindow(t *testing.T) {
```
was bedeutet SinglePlanWindow?

iseeberg79 replied:
renamed to ..._SelectsCheapestContiguousSlots

andig commented on core/planner/planner_continuous.go:
```diff
@@ -0,0 +1,50 @@
+// findContinuousWindow finds the cheapest continuous window of slots for the given duration.
+// - rates are filtered to [now, targetTime] window by caller
+// Returns the selected rates.
+func findContinuousWindow(rates api.Rates, effectiveDuration time.Duration, targetTime time.Time) api.Rates {
```
It's only a short function now- move to helpers or to planner?

andig committed:
wip

andig commented on core/planner/planner_continuous.go:
```diff
@@ -0,0 +1,50 @@
+	var bestIndex *int
+
+	for i := range rates {
+		windowEnd := rates[i].Start.Add(effectiveDuration)
```
Der continuous plan fängt immer "früh" an. Eine Alternative wäre, den hier "spät" anfangen zu lassen indem man den windowStart aus der TargetTime ermittelt?

Hofyyy replied:
In der Bescheibung steht: Chooses the window with the lowest cost. Das ist dann nicht früh sondern das Mathematische Optimum, mit den gegebenen Randbedinungen. Oder meinst du hier denn fall wenn 10 Stunden genau 20 Cent sind und es mehrere Möglichkeiten gibt? Dann würde ich auch eher für spät pledieren.

andig replied:
Verstehe ich nicht- wir fangen doch auch jetzt (wie bei 1h Slots) möglichst spät an zu laden. Das ist ja im Code an anderer Stelle auch noch drin:

```go
			// the first (if not single) slot should start as late as possible
			if IsFirst(slot, plan) && len(plan) > 0 {
				slot.Start = slot.Start.Add(-requiredDuration)
			} else {
				slot.End = slot.End.Add(requiredDuration)
			}
```

Hängt mein Leben nicht dran, aber erscheint mir inkonsistent

iseeberg79 replied:
>  Oder meinst du hier denn fall wenn 10 Stunden genau 20 Cent sind und es mehrere Möglichkeiten gibt? Dann würde ich auch eher für spät pledieren.

Ja, da gibt es natürlich einen Punkt

andig replied:
"früh" heisst am Ende kürzen, "spät" am Anfang- wenn die Summe der Slots nicht der benötigten Ladedauer entspricht. Falls wir uns einig sind ist die Änderung hier trivial, kann ich gerne machen.

andig replied:
Wir reden natürlich nur von +/- 15min

iseeberg79 replied:
Ja, genau: das günstigste+zusammenhängende+späteste Fenster, das möglichst früh startet

Ich hab noch nicht wieder geschaut, sehe gerade nach den anderen Punkten: aus der Erinnerung gibt es aber Tests, die das Verhalten genau so erwarten: wenn ja, würde ich das gern unverändert belassen

Mit Blick auf die letzten Diskussionen zu weniger Ladeleistung am Ende des Ladevorgangs wäre das für mich ein gewolltes, nützliches Verhalten und sollte das alte Verhalten des Planners bewahren für den continuous-Fall.

andig commented on core/planner/planner_continuous.go:
```diff
@@ -0,0 +1,82 @@
+	}
+
+	// Build the best window only once
+	windowEnd := rates[bestStartIndex].Start.Add(effectiveDuration)
```
Hab ich umgebaut, wirf gerne nochmal einen Blick drauf

iseeberg79 commented on core/planner/planner.go:
```diff
@@ -260,5 +238,20 @@ func splitPreconditionSlots(rates api.Rates, preCondStart time.Time) (api.Rates,
 		precond = append(precond, r)
 	}
 
+	// TODO if this is required it lacks a test
```
unter der Annahme vollständiger bzw. lückenloser Tarifdaten ist das ein edge-case: remove

iseeberg79 commented on core/planner/planner.go:
```diff
@@ -192,72 +143,96 @@ func (t *Planner) Plan(requiredDuration, precondition time.Duration, targetTime
-	// reduce target time by precondition duration
-	targetTime = targetTime.Add(-precondition)
-	requiredDuration = max(requiredDuration-precondition, 0)
+	// filter rates to planning window early for performance
```
die Raten werden nur auf das nötige Zeitfenster gekürzt, also Raten vor jetzt und ausserhalb des Planungsfensters entfernt - ich passe den Kommentar an

// clamp rates including precondition window

andig committed:
Consolidate tests

andig committed:
wip

andig committed:
wip

andig committed:
wip

andig committed:
wip
